qiskit/circuit/controlledgate.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Controlled unitary gate.
+"""
+
+from qiskit.exceptions import QiskitError
+from .gate import Gate
+
+
+class ControlledGate(Gate):
+    """Controlled unitary gate."""
+
+    def __init__(self, name, num_qubits, params, label=None, num_ctrl_qubits=1,
+                 definition=None):
```

Comment:
why include `definition` as part of the constructor, and not do it like the base Gate class?
also for the ControlledGate i think it's important to specify it as the `base_gate` and the `num_ctrl_qubits`. here it seems like we give it a name other than controlled-`base_gate`, and give it `params` other than what the base gate already has.

Reply:
Since the __init__ function is frequently used to set object attributes why not set `definition` if it is known? Are you suggesting to instead set `base_gate`? `params` should be the same as the base_gate.

Reply:
I think a controlled-gate should be defined by "the gate that is being controlled" (i.e. `base_gate`) and "the number of controls". So a "controlled-U" gate is completely defined by what U is and by the fact that it is controlled by 1 qubit. So it shouldn't have a new `name` or new `params` -- the `base_gate` already gives us that. 
So I think this can be simplified like this:
```
    def __init__(self, base_gate, num_ctrl_qubits=1, label=None):
```

Here I'm assuming that `base_gate` will have a `.phase` property (to be added later) which changes the global phase of the base_gate, thus changing the resulting controlled-gate.

Reply:
The `q_if(num_ctrl_qubits)` method of all Gates in this PR I think is in the spirit of what you're suggesting. This __init__ function probably wouldn't be used directly as much if someone already had access to a Gate object. This suggestion does change what is thought of as the "definition" of the gate. I interpreted that to actually implement the gate it is defined for instead of the base gate it controls.